#! /bin/sh /usr/share/dpatch/dpatch-run
## postgresql_conf_fix.dpatch by Sebastian Harl <sh@tokkee.org>
##
## DP: postgresql_default.conf:
## DP:
## DP:  * Don't use the deprecated {Min,Max}PgVersion options.
## DP:
## DP:  * Make sure the "disk_io" query does not return NULLs.
## DP:
## DP:    Starting with some version between 8.3.3 and 8.3.6,
## DP:    pg_statio_*_tables returns NULL instead of 0 for statistics if no
## DP:    instance of the appropriate relation exists. PQgetvalue() returns an
## DP:    empty string in that case which would then result in error messages
## DP:    when udb_result_submit() tries to convert that to a number.
## DP:
## DP:    Now, the "disk_io" query uses PostgreSQL's coalesce() function to
## DP:    make sure 0 is returned instead of NULL.

@DPATCH@

diff a/src/postgresql_default.conf b/src/postgresql_default.conf
--- a/src/postgresql_default.conf
+++ b/src/postgresql_default.conf
@@ -54,7 +54,7 @@
 		ValuesFrom "del"
 	</Result>
 
-	MaxPGVersion 80299
+	MaxVersion 80299
 </Query>
 
 <Query queries>
@@ -85,7 +85,7 @@
 		ValuesFrom "hot_upd"
 	</Result>
 
-	MinPGVersion 80300
+	MinVersion 80300
 </Query>
 
 <Query query_plans>
@@ -132,7 +132,7 @@
 		ValuesFrom "dead"
 	</Result>
 
-	MinPGVersion 80300
+	MinVersion 80300
 </Query>
 
 <Query disk_io>
@@ -141,14 +141,14 @@
 </Query>
 
 <Query disk_io>
-	Statement "SELECT sum(heap_blks_read) AS heap_read, \
-			sum(heap_blks_hit) AS heap_hit, \
-			sum(idx_blks_read) AS idx_read, \
-			sum(idx_blks_hit) AS idx_hit, \
-			sum(toast_blks_read) AS toast_read, \
-			sum(toast_blks_hit) AS toast_hit, \
-			sum(tidx_blks_read) AS tidx_read, \
-			sum(tidx_blks_hit) AS tidx_hit \
+	Statement "SELECT coalesce(sum(heap_blks_read), 0) AS heap_read, \
+			coalesce(sum(heap_blks_hit), 0) AS heap_hit, \
+			coalesce(sum(idx_blks_read), 0) AS idx_read, \
+			coalesce(sum(idx_blks_hit), 0) AS idx_hit, \
+			coalesce(sum(toast_blks_read), 0) AS toast_read, \
+			coalesce(sum(toast_blks_hit), 0) AS toast_hit, \
+			coalesce(sum(tidx_blks_read), 0) AS tidx_read, \
+			coalesce(sum(tidx_blks_hit), 0) AS tidx_hit \
 		FROM pg_statio_user_tables;"
 
 	<Result>
